File: Hash Table/0792.Number of Matching Subsequences/Solution.java
class Solution {
    public int numMatchingSubseq(String s, String[] words) {
        List<Deque<String>> ws = new ArrayList<>();
        for (int i = 0; i < 26; i++) {
            ws.add(new ArrayDeque<>());
        }
        for (String w : words) {
            ws.get(w.charAt(0) - 'a').addLast(w);
        }
        int ans = 0;
        for (char c : s.toCharArray()) {
            Deque<String> q = ws.get(c - 'a');
            int size = q.size();
            for (int i = 0; i < size; i++) {
                String w = q.removeFirst();
                if (w.length() == 1) {
                    ans++;
                } else {
                    ws.get(w.charAt(1) - 'a').addLast(w.substring(1));
                }
            }
        }
        return ans;
    }
}

File: Hash Table/0792.Number of Matching Subsequences/Solution.py
class Solution:
    def numMatchingSubseq(self, s: str, words: List[str]) -> int:
        ws = [deque() for _ in range(26)]
        for w in words:
            ws[ord(w[0]) - ord('a')].append(w)
        ans = 0
        for c in s:
            q = ws[ord(c) - ord('a')]
            for _ in range(len(q)):
                v = q.popleft()
                if len(v) == 1:
                    ans += 1
                else:
                    ws[ord(v[1]) - ord('a')].append(v[1:])
        return ans
